flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/TaskManagerLogHandler.java

```diff
@@ -0,0 +1,345 @@
+	private final boolean serveLogFile;
+	private final ExecutionContextExecutor executor;
+
+	private String localJobManagerAddress;
+
+	public TaskManagerLogHandler(
+		JobManagerRetriever retriever,
+		ExecutionContextExecutor executor,
+		Future<String> localJobManagerAddressPromise,
+		FiniteDuration timeout,
+		boolean serveLogFile) throws IOException {
+
+		this(retriever, executor, localJobManagerAddressPromise, timeout, DEFAULT_LOGGER, serveLogFile);
+	}
+
+	public TaskManagerLogHandler(
+		JobManagerRetriever retriever,
+		ExecutionContextExecutor executor,
+		Future<String> localJobManagerAddressFuture,
+		FiniteDuration timeout,
+		Logger logger, boolean serveLogFile) throws IOException {
+
+		this.retriever = checkNotNull(retriever);
+		this.executor = checkNotNull(executor);
+		this.localJobManagerAddressFuture = checkNotNull(localJobManagerAddressFuture);
+		this.timeout = checkNotNull(timeout);
+		this.logger = checkNotNull(logger);
+		this.serveLogFile = serveLogFile;
+	}
+
+	// ------------------------------------------------------------------------
+	//  Responses to requests
+	// ------------------------------------------------------------------------
+
+	@Override
+	public void channelRead0(ChannelHandlerContext ctx, Routed routed) throws Exception {
+		if (localJobManagerAddressFuture.isCompleted()) {
+			if (localJobManagerAddress == null) {
+				localJobManagerAddress = Await.result(localJobManagerAddressFuture, timeout);
+			}
+
+			final HttpRequest request = routed.request();
+
+			Option<Tuple2<ActorGateway, Integer>> jobManager = retriever.getJobManagerGatewayAndWebPort();
+
+			if (jobManager.isDefined()) {
+				// Redirect to leader if necessary
+				String redirectAddress = HandlerRedirectUtils.getRedirectAddress(
+					localJobManagerAddress, jobManager.get());
+
+				if (redirectAddress != null) {
+					HttpResponse redirect = HandlerRedirectUtils.getRedirectResponse(redirectAddress, "");
+					KeepAliveWrite.flush(ctx, routed.request(), redirect);
+				} else {
+					respondAsLeader(ctx, request, routed.pathParams(), jobManager.get()._1());
+				}
+			} else {
+				KeepAliveWrite.flush(ctx, routed.request(), HandlerRedirectUtils.getUnavailableResponse());
+			}
+		} else {
+			KeepAliveWrite.flush(ctx, routed.request(), HandlerRedirectUtils.getUnavailableResponse());
+		}
+	}
+
+	/**
+	 * Response when running with leading JobManager.
+	 */
+	private void respondAsLeader(final ChannelHandlerContext ctx, final HttpRequest request, final Map<String, String> pathParams, final ActorGateway jobManager) throws Exception {
+		final String taskManagerID = pathParams.get(TaskManagersHandler.TASK_MANAGER_ID_KEY);
+
+		boolean fetch;
+		synchronized (lastRequestCompleted) {
+			if (!lastRequestCompleted.containsKey(taskManagerID)) {
+				lastRequestCompleted.put(taskManagerID, true);
+			}
+			fetch = lastRequestCompleted.get(taskManagerID);
+			lastRequestCompleted.put(taskManagerID, false);
+		}
```

Comment:
I think it would be easier to use a `ConcurrentHashMap` in combination with `putIfAbsent` to guard against multiple concurrent requests.